---------------------
PatchSet 340 
Date: 2011/04/20 06:58:54
Author: bluhm
Branch: OPENBSD_4_8
Tag: (none) 
Log:
Perl security fix for CVE-2011-1487.
ok sthen@ jasper@

Members: 
	patchlevel.h:1.17->1.17.4.1 
	pp.c:1.13->1.13.4.1 
	t/op/taint.t:1.12->1.12.4.1 

Index: src/gnu/usr.bin/perl/patchlevel.h
diff -u src/gnu/usr.bin/perl/patchlevel.h:1.17 src/gnu/usr.bin/perl/patchlevel.h:1.17.4.1
--- patchlevel.h:1.17	Mon Oct 12 18:24:21 2009
+++ patchlevel.h	Wed Apr 20 13:58:54 2011
@@ -129,6 +129,7 @@
 #  endif
 static const char * const local_patches[] = {
 	NULL
+	,"CVE-2011-1487"
 	PERL_GIT_UNPUSHED_COMMITS    	/* do not remove this line */
         PERL_GIT_UNCOMMITTED_CHANGES	/* do not remove this line */
 	,NULL
Index: src/gnu/usr.bin/perl/pp.c
diff -u src/gnu/usr.bin/perl/pp.c:1.13 src/gnu/usr.bin/perl/pp.c:1.13.4.1
--- pp.c:1.13	Mon Oct 12 18:24:21 2009
+++ pp.c	Wed Apr 20 13:58:54 2011
@@ -3616,6 +3616,8 @@
 	    SvCUR_set(dest, need - 1);
 	}
     }
+    if (dest != source && SvTAINTED(source))
+	SvTAINT(dest);
     SvSETMAGIC(dest);
     RETURN;
 }
@@ -3719,6 +3721,8 @@
 	    SvCUR_set(dest, d - (U8*)SvPVX_const(dest));
 	}
     }
+    if (dest != source && SvTAINTED(source))
+	SvTAINT(dest);
     SvSETMAGIC(dest);
     RETURN;
 }
@@ -3835,6 +3839,8 @@
 	    SvCUR_set(dest, d - (U8*)SvPVX_const(dest));
 	}
     }
+    if (dest != source && SvTAINTED(source))
+	SvTAINT(dest);
     SvSETMAGIC(dest);
     RETURN;
 }
Index: src/gnu/usr.bin/perl/t/op/taint.t
diff -u src/gnu/usr.bin/perl/t/op/taint.t:1.12 src/gnu/usr.bin/perl/t/op/taint.t:1.12.4.1
--- t/op/taint.t:1.12	Mon Oct 12 18:30:28 2009
+++ t/op/taint.t	Wed Apr 20 13:58:54 2011
@@ -17,7 +17,7 @@
 use File::Spec::Functions;
 
 BEGIN { require './test.pl'; }
-plan tests => 301;
+plan tests => 305;
 
 $| = 1;
 
@@ -1314,6 +1314,19 @@
 
     my $zz = pack "a*a*", q{print "Hello world\n"}, $TAINT;
     ok(tainted($zz), "pack a*a* preserves tainting");
+}
+
+{
+    # [perl #87336] lc/uc(first) failing to taint the returned string
+    my $source = "foo$TAINT";
+    my $dest = lc $source;
+    ok(tainted($dest), "lc(tainted) taints its return value");
+    $dest = lcfirst $source;
+    ok(tainted($dest), "lcfirst(tainted) taints its return value");
+    $dest = uc $source;
+    ok(tainted($dest), "uc(tainted) taints its return value");
+    $dest = ucfirst $source;
+    ok(tainted($dest), "ucfirst(tainted) taints its return value");
 }
 
 # This may bomb out with the alarm signal so keep it last
